[131501519] Checkpoint Review 1

hmasila opened the pull request:
## What does this PR do?
- Creates the shotn application
- A user can login in, signup or shorten their links
- A user can see the top links depending on their number of clicks
- A user can see the recently shortened links
- A user can see the top users
- A logged in user can manage their links by deactivating, activating or deleting them
- A logged in user can view individual details about a link

![image](https://cloud.githubusercontent.com/assets/20906768/19224184/2e054dac-8e89-11e6-9b69-e7a245de7005.png)

_Kindly review the code. Your feedback is highly encouraged_

Commit:
create views

Commit:
feat(views): write views for the shotn application

- the home view
- the dignup page
- the login page
- the popular links panel
- the recent links panel
- sessions for login
- authenticating and authorization

Commit:
feat(controllers): add controllers to control links, sessions and users

add links controller where a user can create a link
 - a registered user can edit and delete a link
add users controller where a user can be saved to the database
 - a user can view their links
add sessions controller where a user can login and logout

add the relevant views for them

Commit:
bg(views): fix the shorten box in the root folder

Commit:
feat(user_links): create a controller that generates user links in the views

- check if vanity string is provided
- check if vanity_string is unique
- create link depending on the logged in user
- create a link even if user is not logged in
- create a vanity string if user did not provide one
-

Commit:
feat(user_links): populate user links or display no links banner

Commit:
Merge branch 'develop' into user-links-list-131432957

Commit:
feat(views): manage user links by updating or deleting them

- create edit page in the links views
- create a forem to edit the links
- create the edit and update actions in the links conntroller
- updated the login and signup forms
- updated the stylesheets to cater for the new pages
- update user links to display how long ago when the link was created

Commit:
Merge pull request #4 from andela-hmasila/manage-links-131484921

[131484921] manage user links by updating or deleting them

Commit:
feat(link details): display link details for logged in users

- create a show page for a partiular link
- create the `show` action in the link controller
- set the link params to the `show` action
- remove the users `show` page and replace it with a link `index` page
- change the `show` action from redirecting `short url` to `original url`. Create an action called `original-url` to do that
- update the routes to include `original-url`

Commit:
Merge pull request #5 from andela-hmasila/view-link-details-131488141

[131488141] display link details for logged in users

Commit:
feat(error page): create an error-inactive page that is rendered when a user clicks a deactivated link

- create `inactive` page
- create `inactive` action in the links controller
- create `inactive` route
- other tasks in this commit
    - moved footer's inline css to the root css
    - modified the `no-links` and links `index` page to be centered
    - renamed the `auth` layout to `plain_layout`
    - updated the sessions and users controllers to use `plain_layout` instead of `auth`

Commit:
Merge pull request #6 from andela-hmasila/create-error-inactive-page-131489495

feat(error page): create an error-inactive page that is rendered when…

Commit:
feat(error-deleted): create an error-deleted page that displays when a user clicks on a deleted link

 - create`error` view
 - create `deleted` partial
 - create controller to set the link
 - create the rout to `/error/:id`
 - other tasks in the commit
    - update the username in the header to redirect to `home_path`

Commit:
Merge pull request #7 from andela-hmasila/create-error-deleted-page-131489573

[131489573] Create error deleted page that displays when a user clicks on a deleted link

Commit:
bug(production): update production to load assets on launch

Commit:
bg(heroku): update procfile for heroku to see ruby version

Commit:
ch(tests): create tests for user model
 - validate presence, length of password
 - validate presence, length and format of name
 - validate presence, format, uniqueness of email
 - ensure encryption and email conversion to downcase  before save
 -ensures a user is created with valid credentials

other tasks in the commit
 - format the root scss
 - remove authentication from user model and put it in session  controller
 - update factories for user models to use valid credentials

Commit:
ch(tests): create tests for link model

- validate presence of full_url and validity of full_url
- validate uniqueness of vanity_string
- ensure vanity string is generated after creation of a link
- ensure most_popular and most_recent do not contain deleted links

Commit:
Merge pull request #8 from andela-hmasila/create-user-model-tests-131112543

[131112543]create user model tests

Commit:
Merge pull request #9 from andela-hmasila/create-link-model-tests-131538959

131538959 create link model tests

Commit:
ch(tests): tests for the users controller
 - returns http success on `#new`
 - given valid attributes
    - creates a new user
    - redirects to `home_path`
 - given invalid attributes
    - does not create a new user
    - redirects to `signup_path`

Commit:
ch(indentation): make code more presemtable by introducing lines between contexts

Commit:
Merge pull request #11 from andela-hmasila/create-user-controller-tests-131546437

[131546437]Create user controller tests

Commit:
ch(tets): add rails-controller-testing gem to allow rendering of templates

Commit:
Merge pull request #12 from andela-hmasila/create-user-controller-tests-131546437

[131546437]Create user controller tests

Commit:
ch(gemfile): add rails_12factor to the production group

Commit:
ch(tests): create tests for the sessions controller

Commit:
bg(views): add loic to check if there is no link in the db

Commit:
bg(views): correct the position of the end_keyword

Commit:
bg(views): use root-url instead of passing hard coded path

Commit:
bg(views): fix flash to display with fading effect

- delete all coffee scripts
- create links.js
- write code to fade out flash
- change flash[:error] to flash[:danger]
- use consistent style for the scss - snake-case and spaces after braces
- freeze constants in constantshelper
- write code to display flash

Commit:
bg(views): fix flashes to appear at the into section

- write code to display flashes on the intro section
- write js code to time the flashes and fade them out
- change flash messages to contain error messages and not only custom strings
- write query to capture top users
- update the top_users views to display the top 5 users
- fix the shortened url flash that appears after a link has been shortened - was only showing home but now shows the shortened link
- put the time logic in the helper

Commit:
bg(views): display copy link after shortening

- install zeroclipboard-rails gem
- create copy link
- create the js function that copies link to the clipboard
- update the css for that copy link
- update deleted to read "You deleted this page"

Commit:
doc(Readme): update readme

Commit:
ch(travis): create the travis file

Commit:
ch(coverals): add coverals to shotn

Commit:
doc(Readme): add coverals to the Readme

Commit:
doc(Readme): add travis badge to the Readme

Commit:
bg(): change the UI while adding tests for controllers

- add a background image
- make site less scrollable by using nav-tabs
- use SecureRandom instead of base 62
- write tests for user model, link model, user controller, sessions controller, root controller and links controller
- add migration for link_count in the User model
- Add migration for title in link model
-Use faker in factorygirl
- use bycrypts encryption of secure_password
- use shoulda to do tests

akabiru commented on app/controllers/root_controller.rb:
```diff
@@ -0,0 +1,8 @@
+    @link = Link.new
+    @top_links = Link.most_popular
+    @recent_links = Link.most_recent
+    @top_users = User.top_users
```
**Code Quality - Nitpick**
@andela-hmasila Consider using the presenter pattern to reduce the number on instance variables in the controller.

``` ruby
def index
  @presenter = Links::IndexPresenter.new
end
```

In `app/presenters/links`

``` ruby
module Links
  class IndexPresenter
    def top_links
      Link.most_popular
    end

    def recent_links
      Link.most_recent
    end
  end
end
```

Thus, in the index view you can do this:

``` erb
<%= @presenter.top_links  %>
<%= @presenter.recent_links  %>
```

makinwab replied:
@andela-hmasila has this been implemented?

akabiru commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,41 @@
+  include ConstantsHelper
+
+  def new
+    render layout: 'plain_layout'
```
@andela-hmasila Are you using a different layout at this point?

hmasila replied:
Yes. It's just a plain layout without the footer and header

akabiru commented on app/helpers/root_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module RootHelper
```
You can delete these zombie files since they're not in use.

akabiru commented on coverage/.last_run.json:
```diff
@@ -0,0 +1,5 @@
+{
```
It would be good to `gitgnore` coverage since there's a badge on the ReadMe.

akabiru commented on spec/controllers/root_controller_spec.rb:
```diff
@@ -0,0 +1,21 @@
+      expect(response).to render_template('index')
+    end
+
+    it 'has valid instance variables' do
```
**Code Quality -Nitpick**
Consider splitting the assertions into their own examples with specific messages. Remember, your tests should be as expressive as possible.

``` ruby
it "creates a new link"
it "assigns top users to @top_users"
it "assigns top links to @top_links"
it "assigns recent links to @recent_links"
```

Also assert the instance variable contains specific objects in addition to  not being nil.

Check this out http://betterspecs.org/

akabiru commented on spec/models/link_spec.rb:
```diff
@@ -0,0 +1,34 @@
+    it { should belong_to(:user) }
+  end
+
+  context 'validations' do
```
Use a `describe` in place of `context`. There are multiple occurrences of this.

http://betterspecs.org/

akabiru commented on spec/views/root/index.html.erb_spec.rb:
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
```
**Code Quality - Critical**
Add feature specs.

https://www.relishapp.com/rspec/rspec-rails/docs/feature-specs/feature-spec

Commit:
bg(tets): update tests to use betterspecs format
- remove zombie files
- use explicit test examples
- write tests for sessioncontroller and make them pass
- remove authenticate method in sessionscontroller
- use bycrypt's method to authenticate users`

Commit:
ch(coverage): remove coverage details from commits

Commit:
feat(tests): add specs for links controller
- add all tests for links controller
- add stubs for login
- update sessions controller and users controller to declare layout before actions
- add title to user_links view

Commit:
bg(views): remove non-url safe characters from vanity strings
- remove non-url safe characters from vanity strings
- sort user links from the most recent
- add public assets
- update font-size of link title

Commit:
ch(tests): add feature tests
- write feature tests for
    - login page
    - signup page
    - home page
    - links - index, show, edit, error page
    - root url page

Commit:
Merge pull request #14 from andela-hmasila/deploy-on-heroku-131501519

[131501519]Deploy on heroku

Commit:
bg(coverage): add badges for travis and coverals to the readme

Commit:
bg(readme): add limitations  to the readme

Commit:
Merge branch 'develop' of github.com:andela-hmasila/shotn into develop

Commit:
ch(readme): add GPA badge to readme

Commit:
ch(readme): add GPA badge to readme

Commit:
bg(UI): make site responsive
- add meta tags to make site responsive on a small screen
- add add col-md to the into section

Commit:
bg(UI/UX): make site more responsive

Commit:
bg(UI/UX): make site more responsive

Commit:
bg(assets): remove duplicate assets

Commit:
ch(tests): use shoulda matchers for model tests

Commit:
ch(tests): use shoulda matchers for link_spec

Commit:
ch(models): use linked scopes

Commit:
ch(models): use linked scopes

Commit:
ch(models): use linked scopes

Commit:
ch(refactor): replace constants helper with a message module
- use methods to define messages in the Messages module
- make links controller thinner by moving the queries to the model
- move methods that cut across to controllers to the application controller

Commit:
bg(login): correct successful login message

Commit:
ch(tests): test links scope in the Link model

Commit:
bg(tets): use unique email to test user links

Commit:
ch(tests): remove chained scopes
- remove chained scopes
- write individual scopes for most_popular and most_recent
- add examples to test most_popular and most recent

Commit:
ch(tests): test features instead of views

Commit:
ch(spec): test inactive and deleted links
- write scenarios for inactive and deleted links
- remove FactoryGirl in `FactoryGirl.attributes_for`
- check for unique emails in FactoryGirl Fakers

Commit:
ch(spec): add whitespaces after end to make code more readable

Commit:
bg(spec): test for link title

Commit:
ch(tests): use more explicit example explanations

makinwab commented on spec/controllers/links_controller_spec.rb:
```diff
@@ -0,0 +1,241 @@
+        }
+      end
+
+      it 'should not increase link count by 1' do
```
This description here is not in line with the expectations you're making.

makinwab commented on spec/controllers/links_controller_spec.rb:
```diff
@@ -0,0 +1,241 @@
+    end
+
+    context 'when link parameters are valid' do
+      it 'updates the link' do
```
You should actually verify that the link has the updated content/value instead of just the success message.

makinwab commented on spec/models/link_spec.rb:
```diff
@@ -0,0 +1,59 @@
+  it { should have_db_column :user_id }
+  it { should have_db_index :user_id }
+
+  describe '#link_title' do
```
You are describing the `link_title` method but are not making your assertions with it.

makinwab commented on spec/models/link_spec.rb:
```diff
@@ -0,0 +1,59 @@
+      expect(Link.most_popular).to_not include(link)
+    end
+
+    it 'should include active_link' do
```
You are not asserting that the link created includes `active_link`.

Please fix.

Commit:
ch(tests): update links_controller tests

Commit:
ch(tests): update links_controller tests

Commit:
bg(tests): rename  to  and
